File: day06/py/solution.py
from pathlib import Path

lines = Path("../input.txt").read_text().splitlines()

def solve(lines):
    width = max(len(line) for line in lines)
    grid = [line.ljust(width) for line in lines]

    last_row = len(grid) - 1
    total = 0

    c = 0
    while c < width:
        if all(grid[r][c] == " " for r in range(len(grid))):
            c += 1
            continue

        start = c
        while c < width and not all(grid[r][c] == " " for r in range(len(grid))):
            c += 1
        end = c

        nums = []
        for r in range(last_row):
            val = grid[r][start:end].strip()
            if val:
                nums.append(int(val))

        op = grid[last_row][start:end].strip()

        if op == "+":
            result = sum(nums)
        elif op == "*":
            result = 1
            for n in nums:
                result *= n

        total += result

    return total

print("Result:", solve(lines))

File: day06/js/solution.js
import fs from "fs";

const lines = fs.readFileSync("../input.txt", "utf-8").trimEnd().split("\n");

function extractPart1(grid, start, end, lastRow) {
  const nums = [];

  for (let r = 0; r < lastRow; r++) {
    const val = grid[r].slice(start, end).trim();
    if (val) nums.push(Number(val));
  }

  return nums;
}

function extractPart2(grid, start, end, lastRow) {
  const nums = [];

  for (let col = end - 1; col >= start; col--) {
    let digits = "";
    for (let row = 0; row < lastRow; row++) {
      const ch = grid[row][col];
      if (ch >= "0" && ch <= "9") {
        digits += ch;
      }
    }
    if (digits) nums.push(Number(digits));
  }

  return nums;
}

function solve(lines, extractNumbers) {
  const width = Math.max(...lines.map((l) => l.length));
  const grid = lines.map((l) => l.padEnd(width, " "));

  const lastRow = grid.length - 1;
  let total = 0;
  let c = 0;

  while (c < width) {
    if (grid.every((row) => row[c] === " ")) {
      c++;
      continue;
    }

    const start = c;
    while (c < width && !grid.every((row) => row[c] === " ")) {
      c++;
    }
    const end = c;

    const nums = extractNumbers(grid, start, end, lastRow);

    const op = grid[lastRow].slice(start, end).trim();

    let result;
    if (op === "+") {
      result = nums.reduce((a, b) => a + b, 0);
    } else {
      result = nums.reduce((a, b) => a * b, 1);
    }

    total += result;
  }

  return total;
}

console.log("Part 1:", solve(lines, extractPart1));
console.log("Part 2:", solve(lines, extractPart2));
